### PR TITLE
add test for reloading lookahead optimizer

### PR DESCRIPTION
**Pull request by BinyanHu**
# Description

Brief Description of the PR: add the test for saving and reloading a Lookahead optimizer. The added test function reveals the bug that the states of a lookahead's inner optimizer are not properly saved and reloaded (#2094). So the test can not pass currently.

## Type of change

- [ ] Bug fix
- [ ] New Tutorial
- [ ] Updated or additional documentation
- [x] Additional Testing
- [ ] New Activation and the changes conform to the [activation contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/activations/README.md#contribution-guidelines)
- [ ] New Callback and the changes conform to the [callback contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/callbacks/README.md#contribution-guidelines)
- [ ] New Image addition and the changes conform to the [image op contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/image/README.md#contribution-guidelines)
- [ ] New Layer and the changes conform to the [layer contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/layers/README.md#contribution-guidelines)
- [ ] New Loss and the changes conform to the [loss contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/losses/README.md#contribution-guidelines)
- [ ] New Metric and the changes conform to the [metric contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/metrics/README.md#contribution-guidelines)
- [ ] New Optimizer and the changes conform to the [optimizer contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/optimizers/README.md#contribution-guidelines)
- [ ] New RNN Cell and the changes conform to the [rnn contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/rnn/README.md#contribution-guidelines)
- [ ] New Seq2seq addition and the changes conform to the [seq2seq contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/seq2seq/README.md#contribution-guidelines)
- [ ] New Text addition and the changes conform to the [text op contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/text/README.md#contribution-guidelines)

# Checklist:

- [x] I've properly [formatted my code according to the guidelines](https://github.com/tensorflow/addons/blob/master/CONTRIBUTING.md#coding-style)
    - [x] By running Black + Flake8
    - [x] By running pre-commit hooks
- [ ] This PR addresses an already submitted issue for TensorFlow Addons
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] This PR contains modifications to C++ custom-ops

# How Has This Been Tested?

If you're adding a bugfix or new feature please describe the tests that you ran to verify your changes:

### CODE REVIEW

**Comment by bhack** [comment]
The original tests mentioned in the ticket was testing model serializzation? 
What are you testing with save and load weights?

**Comment by BinyanHu** [comment]
> 
> 
> The original tests mentioned in the ticket was testing model serializzation?
> What are you testing with save and load weights?

The description from the ticket was inaccurate. I raised that issue and the problem is on save and reloading. Please check the the discussions in the issue(#2094) for more details.

**Comment by bhack** [comment]
I was talking about save and reloading original contributor test in https://github.com/CyberZHG/keras-lookahead/blob/master/tests/test_optimizers.py

**Comment by BinyanHu** [comment]
> 
> 
> I was talking about save and reloading original contributor test in https://github.com/CyberZHG/keras-lookahead/blob/master/tests/test_optimizers.py

That is a different way to save a model. The contributor test uses `model.save` and `keras.models.load_model` to save and load the entire graph.  In our case, we are using `model.save_weights` and `model.load_weights` to save and load only the weights of a model.

**Comment by BinyanHu** [comment]
> 
> 
> I was talking about save and reloading original contributor test in https://github.com/CyberZHG/keras-lookahead/blob/master/tests/test_optimizers.py

Just updated the test file referring to the original contributor test. I added cases of whether the inner optimizer contains slots (adam) or not (sgd) and different saving and reloading methods (weights only and whole graph). All tests of "sgd" pass and those of "adam" fail, which I believe should be enough to demonstrate the problem.

**Comment by bhack** [comment]
Ok, have you double checked https://stackoverflow.com/questions/56805588/does-load-model-load-optimizer-state?

File: tensorflow_addons/optimizers/tests/lookahead_test.py [hunk 1]
```diff
@@ -186,3 +189,68 @@ def test_serialization():
     config = tf.keras.optimizers.serialize(optimizer)
     new_optimizer = tf.keras.optimizers.deserialize(config)
     assert new_optimizer.get_config() == optimizer.get_config()
+
+
+def _init_model(optimizer, init_w):
+    model = tf.keras.models.Sequential()
+    dense = tf.keras.layers.Dense(input_shape=(3,), units=1)
+    model.add(dense)
+    model.compile(Lookahead(optimizer), loss="mse")
```
Comment (bhack) [inline tensorflow_addons/optimizers/tests/lookahead_test.py:198]: Also using only `optimizer` here and removing `_optimizer` recursive check your `test_save_load[cpu-no_weights_only-adam` is failing

Comment (bhack) [reply tensorflow_addons/optimizers/tests/lookahead_test.py:198]: Just to be sure it is not related to https://github.com/tensorflow/tensorflow/issues/40342

Comment (bhack) [reply tensorflow_addons/optimizers/tests/lookahead_test.py:198]: @WindQAQ What do you think?

Comment (BinyanHu) [reply tensorflow_addons/optimizers/tests/lookahead_test.py:198]: image
**Fig.** _name of weights at different stages._

For convenience, I only printed the names of the lookahead optimizer weighs. After model weights reloading, "iter" and weights of the inner optimizer are not created yet. In other cases, the weights are complete.

Also please note that lookahead only collects the weighs of its inner optimizer by a property interface, and that is why we can see the weights of the inner optimizer in `lookahead.weights` but they are not checkpointed: https://github.com/tensorflow/addons/blob/f3a50e7e89e23c3b5049af83cbc3697c3c0a3dcb/tensorflow_addons/optimizers/lookahead.py#L135:#L137

Comment (bhack) [reply tensorflow_addons/optimizers/tests/lookahead_test.py:198]: @qlzh727 Can you expand a little bit over this https://github.com/tensorflow/tensorflow/blame/master/tensorflow/python/keras/engine/training.py#L1986-L1987? Is it just a "preference"?

Comment (bhack) [reply tensorflow_addons/optimizers/tests/lookahead_test.py:198]: I'am asking this cause it seems to work with this formulation (for "weights_only" case):

```
def test_save_load(optimizer, weights_only):
    x = np.random.standard_normal((10000, 3))
    w = np.random.standard_normal((3, 1))
    y = np.dot(x, w) + np.random.standard_normal((10000, 1)) * 1e-4

    init_w = np.random.standard_normal((3, 1))

    model = _init_model(optimizer, init_w)
    model.fit(x, y, epochs=2, shuffle=False)

    with tempfile.TemporaryDirectory() as ckpt_dir:
        new_model = _init_model(optimizer, init_w)
        new_model.fit(x, y, epochs=1, shuffle=False)
        ckpt_path = os.path.join(ckpt_dir, "ckpt")
        if weights_only:
            checkpoint = tf.train.Checkpoint(model=new_model)
            checkpoint.save(file_prefix=ckpt_path)
            new_model = None
            checkpoint.restore(tf.train.latest_checkpoint(ckpt_dir))
            new_model = checkpoint.model

        else:
            new_model.save(ckpt_dir)
            new_model = tf.keras.models.load_model(
                ckpt_dir, custom_objects={"Lookahead": Lookahead}
            )
        new_model.fit(x, y, epochs=1, shuffle=False)
        assert np.allclose(model.predict(x), new_model.predict(x))
```

Comment (bhack) [reply tensorflow_addons/optimizers/tests/lookahead_test.py:198]: Also in `h5` format is working:

```
def test_save_load(optimizer, weights_only):
    np.random.seed(0)
    x = np.random.standard_normal((10000, 3))
    w = np.random.standard_normal((3, 1))
    y = np.dot(x, w) + np.random.standard_normal((10000, 1)) * 1e-4

    init_w = np.random.standard_normal((3, 1))

    model = _init_model(optimizer, init_w)
    model.fit(x, y, epochs=2, shuffle=False)

    with tempfile.TemporaryDirectory() as ckpt_dir:
        new_model = _init_model(optimizer, init_w)
        new_model.fit(x, y, epochs=1, shuffle=False)
        ckpt_path = os.path.join(ckpt_dir, "ckpt")
        if weights_only:
            checkpoint = tf.train.Checkpoint(model=new_model)
            checkpoint.save(file_prefix=ckpt_path)
            del new_model
            checkpoint.restore(tf.train.latest_checkpoint(ckpt_dir))
            new_model = checkpoint.model

        else:
            new_model.save(ckpt_dir + "/test.h5")
            new_model = tf.keras.models.load_model(
                ckpt_dir + "/test.h5"
            )
        new_model.fit(x, y, epochs=1, shuffle=False)
        assert np.allclose(model.predict(x), new_model.predict(x))
```

Comment (BinyanHu) [reply tensorflow_addons/optimizers/tests/lookahead_test.py:198]: Yet it fails if we also recreate the checkpoint. Your code passes because the old model still stays inside `checkpoint`. `h5` saves model weights by serialization, which is correctly handled, so it could pass.
```
def test_save_load(optimizer, weights_only):
    x = np.random.standard_normal((10000, 3))
    w = np.random.standard_normal((3, 1))
    y = np.dot(x, w) + np.random.standard_normal((10000, 1)) * 1e-4

    init_w = np.random.standard_normal((3, 1))

    model = _init_model(optimizer, init_w)
    model.fit(x, y, epochs=2, shuffle=False)

    with tempfile.TemporaryDirectory() as ckpt_dir:
        new_model = _init_model(optimizer, init_w)
        new_model.fit(x, y, epochs=1, shuffle=False)
        ckpt_path = os.path.join(ckpt_dir, "ckpt")
        if weights_only:
            checkpoint = tf.train.Checkpoint(model=new_model)
            checkpoint.save(file_prefix=ckpt_path)
            new_model = _init_model(optimizer, init_w)
            new_checkpoint = tf.train.Checkpoint(model=new_model)
            new_checkpoint.restore(tf.train.latest_checkpoint(ckpt_dir))
        else:
            new_model.save(ckpt_dir)
            new_model = tf.keras.models.load_model(
                ckpt_dir, custom_objects={"Lookahead": Lookahead}
            )
        new_model.fit(x, y, epochs=1, shuffle=False)
        assert np.allclose(model.predict(x), new_model.predict(x))

```

Comment (bhack) [reply tensorflow_addons/optimizers/tests/lookahead_test.py:198]: I suppose that you need to assign after restore?

```
        if weights_only:
            checkpoint = tf.train.Checkpoint(model=new_model)
            checkpoint.save(file_prefix=ckpt_path)
            del checkpoint
            checkpoint = tf.train.Checkpoint(model=new_model)
            checkpoint.restore(tf.train.latest_checkpoint(ckpt_dir))
            new_model = checkpoint.model
```

Comment (BinyanHu) [reply tensorflow_addons/optimizers/tests/lookahead_test.py:198]: `checkpoint.model` is `new_model`:
```
        if weights_only:
            checkpoint = tf.train.Checkpoint(model=new_model)
            checkpoint.save(file_prefix=ckpt_path)
            new_model = _init_model(optimizer, init_w)
            new_checkpoint = tf.train.Checkpoint(model=new_model)
            new_checkpoint.restore(tf.train.latest_checkpoint(ckpt_dir))
            assert new_model is new_checkpoint.model
```

Comment (bhack) [reply tensorflow_addons/optimizers/tests/lookahead_test.py:198]: Ok so it seems that the inner optimizer is not tracked automatically also with explicit ceckpoint API:

```
        if weights_only:
            checkpoint = tf.train.Checkpoint(model=new_model, optimizer=new_model.optimizer._optimizer)
            checkpoint.save(file_prefix=ckpt_path)
            del checkpoint
            del new_model
            new_model = _init_model(optimizer, init_w)
            checkpoint = tf.train.Checkpoint(model=new_model,optimizer=new_model.optimizer._optimizer)
            checkpoint.restore(tf.train.latest_checkpoint(ckpt_dir))
```

Comment (bhack) [reply tensorflow_addons/optimizers/tests/lookahead_test.py:198]: @BinyanHu Can you start to check your `weights_only` case with https://github.com/tensorflow/addons/pull/2126?

Comment (BinyanHu) [reply tensorflow_addons/optimizers/tests/lookahead_test.py:198]: `weights_only=True` passed on that branch, yet `weights_only=False` still fails.

Comment (bhack) [reply tensorflow_addons/optimizers/tests/lookahead_test.py:198]: @BinyanHu Yes for the `SavedModel` case I suspect that this is still vaild https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/saving/saved_model/load.py#L99-L104 (especially on optimizers with slots).

Comment (BinyanHu) [reply tensorflow_addons/optimizers/tests/lookahead_test.py:198]: Sure. So after the tracking of inner optimizer weights is fixed, the remaining parts should be tensorflow issues. Besides, similar wrapper optimizers such as `MovingAverage` should also be fixed in this way. Anything left to do with this pull request?

Comment (bhack) [reply tensorflow_addons/optimizers/tests/lookahead_test.py:198]: Yes `AvarageWrapper` was added.

**Comment by bhack** [comment]
@WindQAQ Let me know if you still find useful to have a save_weight/load_weight test. In that case give a pass over the test.
